src/validations/html/rules/MismatchedTag.js

```diff
@@ -0,0 +1,139 @@
+import {List, Map, Record, Set} from 'immutable';
+import Code from '../rules/Code';
+
+const Location = Record({row: 0, column: 0});  // eslint-disable-line new-cap
+
+export default class {
```

Comment:
Unfortunately, I think the problem is that the misplaced close tag may appear _before_ its matching open tag.

In other words:

> Any time we encounter a mismatch:
> - If the closing tag we have encountered does not match an opening tag in the stack

...but the close tag may match a _later_ open tag in the stream of tags.

I'm not opposed to three collections that each correspond to one of the three errors, but how could they be used while avoiding this issue?

While drafting this PR, I initially tried a similar approach until I realized this issue, hence the more complex containers.